designs/v1-api.md

```diff
@@ -0,0 +1,535 @@
+
+1. Apply the updated NodePool, NodeClaim, and EC2NodeClass CRDs, which will contain a `v1` versions listed under the `versions` section of the CustomResourceDefinition
+2. Upgrade Karpenter controller to its `v1` version. This version of Karpenter will start reasoning in terms of the `v1` API schema in its API requests. Resources will be converted from the v1beta1 to the v1 version at runtime, using conversion webhooks shipped by the upstream Karpenter project and the Cloud Providers (for NodeClass changes).
+3. Users update their `v1beta1` manifests that they are applying through IaC or GitOps to use the new `v1` version.
+4. Karpenter drops the `v1beta1` version from the `CustomResourceDefinition` and the conversion webhooks on the next minor version release of Karpenter, leaving no webhooks and only the `v1` version still present
+
+## NodePool API
+
+```
+apiVersion: karpenter.sh/v1
+kind: NodePool
+metadata:
+  name: default
+spec:
+  template:
+    metadata:
+      labels:
+        billing-team: my-team
+      annotations:
+        example.com/owner: "my-team"
+    spec:
+      nodeClassRef:
+        group: karpenter.k8s.aws # Updated since only a single version will be served
+        kind: EC2NodeClass
+        name: default
+      taints:
+        - key: example.com/special-taint
+          effect: NoSchedule
+      startupTaints:
+        - key: example.com/another-taint
+          effect: NoSchedule
+      requirements:
+        - key: "karpenter.k8s.aws/instance-category"
+          operator: In
+          values: ["c", "m", "r"]
+          minValues: 2 # Alpha field, added for spot best practices support
+  disruption:
+    budgets:
+      - nodes: 0
+        schedule: "0 10 * * mon-fri"
+        duration: 16h
+        reasons:
+          - drifted
+          - expired
+      - nodes: 100%
+        reasons:
+          - empty
+      - nodes: "10%"
+      - nodes: 5
+    expireAfter: 720h | Never
+    consolidationPolicy: WhenUnderutilized | WhenEmpty
+    consolidateAfter: 1m | Never # Added to allow additional control over consolidation aggressiveness
+  weight: 10
+  limits:
+    cpu: "1000"
+    memory: 1000Gi
+status:
+  conditions:
+    - lastTransitionTime: "2024-02-02T19:54:34Z"
+      status: "True"
+      type: NodeClassReady
+    - lastTransitionTime: "2024-02-02T19:54:34Z"
+      status: "True"
+      type: Ready
+  resources:
+    cpu: "20"
+    memory: "8192Mi"
+    ephemeral-storage: "100Gi"
+```
+
+### Defaults
+
+Defaults are unchanged in the v1 API version but are included in the API proposal for completeness.
+
+`disruption.budgets` : `{"nodes": "10%"}`
+`disruption.expireAfter`: `30d/720h`
+`disruption.consolidationPolicy`: `WhenUnderutilized`
+`disruption.consolidateAfter`: `15s`
```

Comment:
I don't see how these can be the current defaults when `consolidateAfter` doesn't yet work with `WhenUnderutilized` (which needs #1218)

Reply:
^ +1

Reply:
Good call. I'm dropping this line. Added by mistake.